src/mds/Beacon.cc

```diff
@@ -173,7 +173,11 @@ void Beacon::send_and_wait(const double duration)
   while (!seq_stamp.empty() && seq_stamp.begin()->first <= awaiting_seq) {
     auto now = clock::now();
     auto s = duration*.95-std::chrono::duration<double>(now-start).count();
-    if (s < 0) break;
+    if (s < 0) {
```

Comment:
I was checking the beacon ack code once again, before writing it's test case, and I think this might be the wrong place to decide that we've a 'missed beacon ack' event. Because this comes before the `cvar.wait_for()` wait call. May be [1] might be the right place be to considered as the event happened? Another method would be, we need to timeout the wait and conclude that we have missed the beacon ack.  @vshankar @lxbsz What's your opinion?

[1] https://github.com/ceph/ceph/blob/master/src/mds/Beacon.cc#L151.

Reply:
I don't think the code is a problem, because it's checking the passed time inside a loop. The first time we may not actually have been waiting and so it's a bit silly, but then it works the rest of the time because we wait on the condition variable (with a wakeup happening at the timeout if nothing else signals it).

Reply:
Agree with @gregsfortytwo and this should be okay.

Reply:
@gregsfortytwo That's right. I read the code again and recollect now why I did this change. Thanks :)

Reply:
@gregsfortytwo @lxbsz 

Another question regarding this change is:
These [1][2] are the two callers of `send_and_wait`. Are these the only events (Rank damaged and Daemon Suicide) the 'missed beacon ack' could happen? What if these two events never happen?

[1] https://github.com/ceph/ceph/blob/master/src/mds/MDSRank.cc#L941
[2] https://github.com/ceph/ceph/blob/master/src/mds/MDSDaemon.cc#L827

Reply:
@batrick 
Why do you always timeout [here](https://github.com/ceph/ceph/blob/a5fc29b95281c6ca58c9177c665c379846beb4b3/src/mds/Beacon.cc#L86) ? If it was waiting for an ack like `cvar.wait_for()` does, then I could've consider the timeout as 'missed beacon ack' event. Any reason why it's not doing so?

Reply:
> @batrick Why do you always timeout [here](https://github.com/ceph/ceph/blob/a5fc29b95281c6ca58c9177c665c379846beb4b3/src/mds/Beacon.cc#L86) ? If it was waiting for an ack like `cvar.wait_for()` does, then I could've consider the timeout as 'missed beacon ack' event. Any reason why it's not doing so?

IMO there has another place will call the `_send()`, not only in this thread. Just wait for 1s instead here we could detect that and defer the next `_send()`.